ui/src/dropdown.tsx

```diff
@@ -152,119 +169,160 @@ const
+      [items, setItems, onSearchChange] = useItems(choices, model.value),
+      toggleDialog = () => setIsDialogHidden(v => !v),
+      onChecked = (itemName: S) => (_ev?: React.FormEvent<HTMLElement | HTMLInputElement>, checked = false) => {
+        setItems(items => items.map(item => (({...item, checked: item.name === itemName ? checked : false, show: true}))))
+        wave.args[name] = itemName
+        model.value = itemName
+        if (trigger) wave.push()
         toggleDialog()
-        setFilteredItems(items.map(i => { i.checked = initialSelectedMap.has(i.name); return i }))
-      }, [initialSelectedMap, items, toggleDialog]),
-      submit = React.useCallback((checkedItem?: DropdownItem) => {
-        const result = checkedItem ? [checkedItem] : items.filter(({ checked }) => checked)
-        wave.args[name] = result.length === 1 ? result[0].name : result.map(({ name }) => name)
-
+      }
+    
+      React.useEffect(() => {
+        if (model.value !== undefined) {
+          wave.args[name] = model.value ?? null
+          setItems(items => items.map(i => ({ ...i, checked: model.value === i.name })))
+        }
+      }, [name, model.value, setItems])
+    
+      return (
+        <>
+          <Fluent.TextField
+            data-test={name}
+            iconProps={{ iconName: 'ChevronDown' }}
+            readOnly
+            onClick={toggleDialog}
+            label={label}
+            disabled={disabled}
+            required={required}
+            styles={{ field: { cursor: 'pointer' }, icon: { fontSize: 12, color: cssVar('$neutralSecondary') } }}
+            placeholder={placeholder}
+            value={items.find(i => i.checked)?.text ?? ''} />
+          <Fluent.Dialog hidden={isDialogHidden} dialogContentProps={{ title: label, type: Fluent.DialogType.close }} onDismiss={toggleDialog} minWidth={600} maxWidth='90vw'>
+            <Fluent.DialogContent styles={{ innerContent: { height: '65vh' }, header: { height: 0 } }}>
+              <Fluent.Label>Search</Fluent.Label>
+              <Fluent.SearchBox data-test={`${name}-search`} onChange={onSearchChange} autoFocus />
+              <Fluent.ScrollablePane styles={{ root: { marginTop: 100, boxShadow: `0px 3px 7px ${cssVar('$text3')}` } }}>
+                <Fluent.List
+                  items={items.filter(i => i.show)}
+                  getPageSpecification={getPageSpecification}
+                  renderedWindowsAhead={3}
+                  onRenderCell={onRenderCell(onChecked)}
+                />
+              </Fluent.ScrollablePane>
+            </Fluent.DialogContent>
+          </Fluent.Dialog>
+        </>
+      )
+  },
+  DialogDropdownMulti = ({ name, choices = [], values = [], disabled, required, trigger, placeholder, label }: Dropdown) => {
+    const
+      [isDialogHidden, setIsDialogHidden] = React.useState(true),
+      [items, setItems, onSearchChange] = useItems(choices, values),
+      auxItems = React.useRef([...items]),
```

Comment:
Wouldn't this be enough? Also naming needs work. Something like initial selection or something. Aux is too general.

```suggestion
      auxItems = React.useRef(items),
```

Reply:
> Wouldn't this be enough? 

It's enough because I am always creating a new array by using `map` in `setItems` and `auxItems`.
But semantically, `auxItems` is a copy so it makes sense to create a new array so that in the case someone (by mistake or not) mutates `items`, it won't be affected.

In terms of memory, I profiled `useRef(items)` and `useRef([...items])` and for some reason `useRef([...items])`  consistently performed slightly better. I don't know if using Chrome Devtools is enough to evaluate such a situation though.

3 and 7 are `useRef(items)` the rest is `useRef([...items])`.
<img width="199" alt="image" src="https://user-images.githubusercontent.com/15820796/192238989-7c50d82c-03db-4d45-9317-33ddb9bf7ea4.png">


>Also naming needs work. Something like initial selection or something. Aux is too general.

True. I'd not use `initialSelection` because it can imply that was it the first selection when the component was created. I'm going to use `itemsBackup` since it's used to revert to a previous state.

Reply:
> so that in the case someone (by mistake or not) mutates items, it won't be affected.

This is a shallow copy, not a deep copy so this doesn't hold true - it should be affected in both scenarios.

> I don't know if using Chrome Devtools is enough to evaluate such a situation though.

AFAIK it should allow you to explore the complete memory snapshot item by item so you can see what's the root cause of the extra memory.

> it can imply that was it the first selection when the component was created. I'm going to use itemsBackup since it's used to revert to a previous state

What is previous state in this case? I thought it's the initial selection that is seen after opening the dialog, but maybe I am wrong?

Reply:
> This is a shallow copy, not a deep copy so this doesn't hold true - it should be affected in both scenarios.

Good point, I overlooked that. I guess we can simply pass the `items` reference.

> AFAIK it should allow you to explore the complete memory snapshot item by item so you can see what's the root cause of the extra memory.

If I wanted a detailed explanation, yes, but in this case, I just want to see the overall change to memory since the only thing changed in my code was `useRef([...items])` -> `useRef(items)`. I'm not very familiar with the profiler tool though.

> What is previous state in this case? I thought it's the initial selection that is seen after opening the dialog, but maybe I am wrong?

You are right, you are saying that it's "initial" at the moment we open the dialog. I'm saying that the word "initial" can also mean "when the component is initialized and these values will not change" and that can be confusing. Maybe rename to `snapshot`?
I believe that anyone will be able to follow the logic anyway. I'm fine with `initialSelection` as well.

Reply:
>  just want to see the overall change to memory since the only thing changed in my code was useRef([...items]) -> useRef(items)

I see. I was just wondering about the cause because IMO it should be the other way around - using reference should result in less memory.

> when the component is initialized and these values will not change" and that can be confusing

Got it. `snapshot` is still too broad. Maybe something in the lines of `itemsOnOpen` / `snapshotOnOpen` or maybe even `onDialogOpen` which is the most crystal clear possible probably :D

Reply:
`onDialogOpen` sounds like an event handler. I went with `itemsOnDialogOpen`.